applications/OptimizationApplication/custom_utilities/control/sigmoidal_projection_utils.cpp

```diff
@@ -0,0 +1,277 @@
+
+bool HasVectorDuplicates(const std::vector<double>& rValues)
+{
+    std::set<double> unique_values(rValues.begin(), rValues.end());
+    return unique_values.size() != rValues.size();
+};
+
+void CheckXYVectors(
+    const std::vector<double>& rXValues,
+    const std::vector<double>& rYValues)
+{
+    KRATOS_ERROR_IF(rXValues.size() != rYValues.size())
+        << "SigmoidalProjectionUtils: rXLimits and rYLimits should have the same size.\n";
+
+    KRATOS_ERROR_IF(rXValues.size() < 2)
+        << "SigmoidalProjectionUtils: rXLimits and rYLimits should have at least two entries.\n";
+
+    KRATOS_ERROR_IF(!std::is_sorted(rXValues.begin(), rXValues.end()))
+        << "SigmoidalProjectionUtils: rXLimits should be sorted ascending.\n";
+
+    KRATOS_ERROR_IF(!std::is_sorted(rYValues.begin(), rYValues.end()))
+        << "SigmoidalProjectionUtils: rYValues should be sorted ascending.\n";
+
+    KRATOS_ERROR_IF(HasVectorDuplicates(rXValues))
+        << "SigmoidalProjectionUtils: rXValues have duplications.\n";
+
+    KRATOS_ERROR_IF(HasVectorDuplicates(rYValues))
+        << "SigmoidalProjectionUtils: rYValues have duplications.\n";
+}
+
+IndexType GetUpperValueRangeIndex(
+    const double Value,
+    const std::vector<double>& rRanges)
+{
+    const  IndexType size = rRanges.size();
+    IndexType upper_index;
+    for (upper_index = 0; upper_index < size; ++upper_index) {
+        if (Value < rRanges[upper_index]) {
+            break;
+        }
+    }
+
+    return std::clamp(upper_index, 1UL, size - 1);
+}
+
+double ProjectValueForward(
+    const double xValue,
+    const std::vector<double>& rXLimits,
+    const std::vector<double>& rYLimits,
+    const double Beta,
+    const int PenaltyFactor)
+{
+    const IndexType upper_index = GetUpperValueRangeIndex(xValue, rXLimits);
+
+    const double x1 = rXLimits[upper_index - 1];
+    const double x2 = rXLimits[upper_index];
+    const double y1 = rYLimits[upper_index - 1];
+    const double y2 = rYLimits[upper_index];
+
+    const double limit = std::log1p(std::numeric_limits<double>::max());
+    double pow_val = -2.0 * Beta * (xValue - (x1 + x2) / 2)  * PenaltyFactor;
+    pow_val = std::clamp(pow_val, -limit, limit);
+    return (y2 - y1) / std::pow((1.0 + std::exp(pow_val)), PenaltyFactor) + y1;
+}
+
+double ProjectValueBackward(
+    const double yValue,
+    const std::vector<double>& rXLimits,
+    const std::vector<double>& rYLimits,
+    const double Beta,
+    const int PenaltyFactor)
+{
+    IndexType size = rXLimits.size();
+    KRATOS_ERROR_IF(yValue > rYLimits[size - 1] || yValue < rYLimits[0])
+        << "SigmoidalProjectionUtils::ProjectValueBackward: yValue "
+        << yValue << " is out of the given range " << rYLimits << "\n";
+
+    KRATOS_ERROR_IF(std::abs(yValue)>std::numeric_limits<double>::max())
+        << "SigmoidalProjectionUtils::ProjectValueBackward: yValue "
+        << yValue << " is too big for backward projection !!! " << rYLimits << "\n";
```

Comment:
What is the purpose of this check? In the line 103 it is checked against the upper bounds of the `rYLimits`, hence unless user probides a `NaN` (which is impossible via json), this check is redundant. Am I missing something?